File: app/src/main/java/com/kimjisub/launchpad/BaseActivity.java
package com.kimjisub.launchpad;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.graphics.drawable.Drawable;
import android.media.AudioManager;
import android.os.Bundle;
import android.os.Process;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.util.DisplayMetrics;
import android.widget.LinearLayout;
import android.widget.Toast;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.InterstitialAd;
import com.kimjisub.launchpad.manager.FileManager;
import com.kimjisub.launchpad.manager.Log;
import com.kimjisub.launchpad.manager.PreferenceManager;
import com.vungle.warren.InitCallback;
import com.vungle.warren.Vungle;

import java.io.File;
import java.util.ArrayList;

import static com.kimjisub.launchpad.manager.Constant.ADMOB;
import static com.kimjisub.launchpad.manager.Constant.ADSCOOLTIME;
import static com.kimjisub.launchpad.manager.Constant.VUNGLE;

public abstract class BaseActivity extends AppCompatActivity {

	private static InterstitialAd interstitialAd;

	public File F_UniPackRootExt;
	public File F_UniPackRootInt;

	public static int Scale_Width = 0;
	public static int Scale_Height = 0;
	public static int Scale_PaddingWidth = 0;
	public static int Scale_PaddingHeight = 0;

	private void initVar() {
		F_UniPackRootExt = FileManager.getExternalUniPackRoot();
		F_UniPackRootInt = FileManager.getInternalUniPackRoot(getApplicationContext());

		FileManager.makeDirWhenNotExist(F_UniPackRootExt);
		FileManager.makeNomedia(F_UniPackRootExt);
	}

	// ============================================================================================= Data binding

	<T extends ViewDataBinding> T setContentViewBind(int layoutId) {
		return DataBindingUtil.setContentView(this, layoutId);
	}

	// ============================================================================================= Activity

	public static ArrayList<Activity> activityList = new ArrayList<>();

	static void startActivity(Activity activity) {
		activityList.add(activity);
		printActivityLog(activity.getLocalClassName() + " start");
	}

	static void finishActivity(Activity activity) {
		boolean exist = false;
		int size = activityList.size();
		for (int i = 0; i < size; i++) {
			if (activityList.get(i) == activity) {
				activityList.get(i).finish();
				activityList.remove(i);
				exist = true;
				break;
			}
		}
		printActivityLog(activity.getLocalClassName() + " finish" + (exist ? "" : " error"));
	}

	static void restartApp(Activity activity) {

		int size = activityList.size();
		for (int i = size - 1; i >= 0; i--) {
			activityList.get(i).finish();
			activityList.remove(i);
		}
		activity.startActivity(new Intent(activity, MainActivity.class));
		printActivityLog(activity.getLocalClassName() + " requestRestart");

		Process.killProcess(Process.myPid());
	}

	static void printActivityLog(String log) {
		StringBuilder str = new StringBuilder("ACTIVITY STACK - " + log + "[");
		int size = activityList.size();
		for (int i = 0; i < size; i++) {
			Activity activity = activityList.get(i);
			str.append(", ").append(activity.getLocalClassName());
		}
		Log.activity(str + "]");
	}

	static void requestRestart(final Context context) {
		new AlertDialog.Builder(context)
				.setTitle(lang(context, R.string.requireRestart))
				.setMessage(lang(context, R.string.doYouWantToRestartApp))
				.setPositiveButton(lang(context, R.string.restart), (dialog, which) -> {
					restartApp((Activity) context);
					dialog.dismiss();
				})
				.setNegativeButton(lang(context, R.string.cancel), (dialog, which) -> {
					dialog.dismiss();
					((Activity) context).finish();
				})
				.show();
	}

	// ============================================================================================= Function

	void rescanScale(LinearLayout LL_scale, LinearLayout LL_paddingScale) {
		Scale_Width = LL_scale.getWidth();
		Scale_Height = LL_scale.getHeight();
		Scale_PaddingWidth = LL_paddingScale.getWidth();
		Scale_PaddingHeight = LL_paddingScale.getHeight();
	}


	File[] getUniPackDirList() {
		File[] projectFiles = FileManager.addFileArray(F_UniPackRootExt.listFiles(), F_UniPackRootInt.listFiles());
		File[] projectFilesSort = FileManager.sortByTime(projectFiles);

		return projectFilesSort;
	}

	// ============================================================================================= Ads

	public boolean checkAdsCooltime() {
		long prevTime = PreferenceManager.PrevAdsShowTime.load(this);
		long currTime = System.currentTimeMillis();

		return (currTime < prevTime) || currTime - prevTime >= ADSCOOLTIME;
	}

	public void updateAdsCooltime() {
		long currTime = System.currentTimeMillis();
		PreferenceManager.PrevAdsShowTime.save(this, currTime);
	}

	public void initVungle() {
		if (!Vungle.isInitialized()) {
			Log.vungle("isInitialized() == false");
			Log.vungle("init start");
			Vungle.init(VUNGLE.APPID, getApplicationContext(), new InitCallback() {
				@Override
				public void onSuccess() {
					// Initialization has succeeded and SDK is ready to load an ad or play one if there
					// is one pre-cached already
					Log.vungle("init onSuccess()");
				}

				@Override
				public void onError(Throwable throwable) {
					// Initialization error occurred - throwable.getLocalizedMessage() contains error message
					Log.vungle("init onError() == " + throwable.getLocalizedMessage());
				}

				@Override
				public void onAutoCacheAdAvailable(String placementId) {
					// Callback to notify when an ad becomes available for the auto-cached placement
					// NOTE: This callback works only for the auto-cached placement. Otherwise, please use
					// LoadAdCallback with loadAd API for loading placements.
					Log.vungle("init onAutoCacheAdAvailable()");
				}
			});
		} else {
			Log.vungle("isInitialized() == true");
		}
	}

	public void showAdmob() {
		Log.admob("showAdmob ================================");

		boolean isLoaded = interstitialAd.isLoaded();
		Log.admob("isLoaded: " + isLoaded);

		if (isLoaded) {
			interstitialAd.show();
			loadAdmob();
			Log.admob("show!");
		} else {
			interstitialAd.setAdListener(new AdListener() {
				public void onAdLoaded() {
					interstitialAd.show();
					loadAdmob();
					Log.admob("show!");
				}
			});
		}
	}

	void loadAdmob() {
		Log.admob("loadAdmob ================================");
		interstitialAd = new InterstitialAd(this);
		interstitialAd.setAdUnitId(ADMOB.MAIN_START);
		interstitialAd.loadAd(new AdRequest.Builder()
				.addTestDevice("36C3684AAD25CDF5A6360640B20DC084")
				.build());
	}

	// ============================================================================================= Show Things, Get Resources


	public void showToast(String msg) {
		showToast(this, msg);
	}

	public void showToast(int resId) {
		showToast(this, resId);
	}

	public static void showToast(Context context, String msg) {
		Toast.makeText(context, msg, Toast.LENGTH_SHORT).show();
	}

	public static void showToast(Context context, int resId) {
		Toast.makeText(context, resId, Toast.LENGTH_SHORT).show();
	}

	public void showDialog(String title, String content) {
		showDialog(this, title, content);
	}

	public static void showDialog(Context context, String title, String content) {
		new AlertDialog.Builder(context)
				.setTitle(title)
				.setMessage(content)
				.setPositiveButton(lang(context, R.string.accept), null)
				.show();
	}

	public String lang(int id) {
		return lang(this, id);
	}

	public static String lang(Context context, int id) {
		return context.getResources().getString(id);
	}

	public int color(int id) {
		return color(this, id);
	}

	public static int color(Context context, int id) {
		return context.getResources().getColor(id);
	}

	public Drawable drawable(int id) {
		return drawable(this, id);
	}

	public static Drawable drawable(Context context, int id) {
		return context.getResources().getDrawable(id);
	}

	public int dpToPx(float dp) {
		DisplayMetrics metrics = getResources().getDisplayMetrics();
		float px = dp * (metrics.densityDpi / 160f);
		return Math.round(px);
	}

	public int pxToDp(int pixel) {
		float dp = 0;
		try {
			DisplayMetrics metrics = getResources().getDisplayMetrics();
			dp = pixel / (metrics.densityDpi / 160f);
		} catch (Exception ignored) {
		}
		return (int) dp;
	}

	// ============================================================================================= Activity Cycle

	@Override
	public void onCreate(Bundle savedInstanceState) {
		Log.activity("onCreate " + this.getLocalClassName());
		super.onCreate(savedInstanceState);
		initVar();

		startActivity(this);
	}

	@Override
	public void onStart() {
		Log.activity("onStart " + this.getLocalClassName());
		super.onStart();
	}

	@Override
	public void onResume() {
		Log.activity("onResume " + this.getLocalClassName());
		super.onResume();
		initVar();

		this.setVolumeControlStream(AudioManager.STREAM_MUSIC);
		initVungle();
	}

	@Override
	public void onPause() {
		Log.activity("onPause " + this.getLocalClassName());
		super.onPause();
	}

	@Override
	public void onStop() {
		Log.activity("onStop " + this.getLocalClassName());
		super.onStop();
	}

	@Override
	public void onRestart() {
		Log.activity("onRestart " + this.getLocalClassName());
		super.onRestart();
	}

	@Override
	public void onDestroy() {
		Log.activity("onDestroy " + this.getLocalClassName());
		super.onDestroy();
		finishActivity(this);
	}
}

File: app/src/main/java/com/kimjisub/launchpad/activity/BaseActivity.kt
package com.kimjisub.launchpad.activity

import android.app.Activity
import android.app.Application
import android.content.DialogInterface
import android.content.Intent
import android.media.AudioManager
import android.os.Build
import android.os.Bundle
import android.os.Process
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.kimjisub.launchpad.R.anim
import com.kimjisub.launchpad.R.string
import com.kimjisub.launchpad.db.repository.UnipackRepository
import com.kimjisub.launchpad.manager.PreferenceManager
import com.kimjisub.launchpad.manager.WorkspaceManager
import com.kimjisub.launchpad.tool.Log
import org.koin.android.ext.android.inject
import splitties.activities.start

open class BaseActivity : AppCompatActivity() {

	companion object {
		private val activities = mutableListOf<BaseActivity>()

		private val lifecycleCallbacks = object : Application.ActivityLifecycleCallbacks {
			override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
				if (activity is BaseActivity) {
					activities.add(activity)
					printActivityLog("${activity.getActivityName()} start")
				}
			}

			override fun onActivityDestroyed(activity: Activity) {
				if (activity is BaseActivity) {
					val removed = activities.remove(activity)
					printActivityLog("${activity.getActivityName()} finish${if (removed) "" else " error"}")
				}
			}

			override fun onActivityStarted(activity: Activity) {}
			override fun onActivityResumed(activity: Activity) {}
			override fun onActivityPaused(activity: Activity) {}
			override fun onActivityStopped(activity: Activity) {}
			override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
		}

		private var callbacksRegistered = false

		internal fun registerCallbacks(application: Application) {
			if (!callbacksRegistered) {
				application.registerActivityLifecycleCallbacks(lifecycleCallbacks)
				callbacksRegistered = true
			}
		}

		private fun restartApp(activity: BaseActivity) {
			for (a in activities.asReversed()) {
				a.finish()
			}
			activities.clear()
			activity.start<MainActivity>()
			printActivityLog("${activity.getActivityName()} requestRestart")
			Process.killProcess(Process.myPid())
		}

		private fun printActivityLog(log: String) {
			val stack = activities.joinToString(", ") { it.getActivityName() }
			Log.activity("ACTIVITY STACK - $log[$stack]")
		}

		fun requestRestart(activity: BaseActivity) {
			AlertDialog.Builder(activity)
				.setTitle(activity.getString(string.requireRestart))
				.setMessage(activity.getString(string.doYouWantToRestartApp))
				.setPositiveButton(activity.getString(string.restart)) { dialog: DialogInterface, _: Int ->
					restartApp(activity)
					dialog.dismiss()
				}
				.setNegativeButton(activity.getString(string.cancel)) { dialog: DialogInterface, _: Int ->
					dialog.dismiss()
					activity.finish()
				}
				.show()
		}
	}

	val p: PreferenceManager by inject()
	val ws: WorkspaceManager by inject()
	val unipackRepo: UnipackRepository by inject()

	fun getActivityName(): String {
		return localClassName.split('.').last()
	}

	override fun startActivity(intent: Intent) {
		super.startActivity(intent)
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
			overrideActivityTransition(OVERRIDE_TRANSITION_OPEN, anim.activity_in, anim.activity_out)
		} else {
			@Suppress("DEPRECATION")
			overridePendingTransition(anim.activity_in, anim.activity_out)
		}
	}

	override fun onCreate(savedInstanceState: Bundle?) {
		Log.activity("onCreate ${getActivityName()}")
		super.onCreate(savedInstanceState)

		registerCallbacks(application)
	}

	override fun onStart() {
		Log.activity("onStart ${getActivityName()}")
		super.onStart()
	}

	override fun onResume() {
		Log.activity("onResume ${getActivityName()}")
		super.onResume()
		volumeControlStream = AudioManager.STREAM_MUSIC
	}

	override fun onPause() {
		Log.activity("onPause ${getActivityName()}")
		super.onPause()
	}

	override fun onStop() {
		Log.activity("onStop ${getActivityName()}")
		super.onStop()
	}

	override fun onRestart() {
		Log.activity("onRestart ${getActivityName()}")
		super.onRestart()
	}

	override fun onDestroy() {
		Log.activity("onDestroy ${getActivityName()}")
		super.onDestroy()
	}
}